File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/cache/operation/OutOfBandPayloadDecodeOperation.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.cache.operation

import ch.nevis.mobile.authentication.sdk.react.cache.state.NoOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.OperationState
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.userinteraction.onerror.DefaultOnErrorImpl
import ch.nevis.mobile.authentication.sdk.react.userinteraction.onsuccess.DefaultOnSuccessImpl
import ch.nevis.mobile.sdk.api.MobileAuthenticationClientError
import ch.nevis.mobile.sdk.api.operation.outofband.OutOfBandPayload

data class OutOfBandPayloadDecodeOperation(
    override var state: OperationState,
    override var onSuccess: (OutOfBandPayload?) -> Unit,
    override var onError: (MobileAuthenticationClientError) -> Unit,
) : Operation<OutOfBandPayload>() {
    constructor(operationId: String) : this(
        onSuccess = {
            DefaultOnSuccessImpl(operationId, ReactMethodName.OOB_PAYLOAD_DECODE).onSuccess(outOfBandPayload = it)
        },
        onError = {
            DefaultOnErrorImpl(operationId, ReactMethodName.OOB_PAYLOAD_DECODE).onError(it)
        },
        state = NoOperation
    )

    override fun cancel() {
        state.cancel()
    }
}

File: ios/Cache/Operation/OutOfBandPayloadDecodeOperation.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

class OutOfBandPayloadDecodeOperation: Operation {
	// MARK: Properties

	var state: OperationState
	var onError: ErrorHandler
	let onSuccess: (OutOfBandPayload) -> ()

	// MARK: Initialization

	init(operationId: String) {
		self.state = NoOperationState()
		self.onSuccess = {
			OnSuccessImpl(operationId: operationId, method: .payloadDecode).onSuccess(payload: $0)
		}

		self.onError = { OnErrorImpl(operationId: operationId, method: .payloadDecode).onError($0) }
	}

	// MARK: Operation

	func cancel() {
		state.cancel()
	}
}
